### PR TITLE
pc-hosted semihosting

### PR DESCRIPTION
**Pull request by koendv**
As requested, semihosting for "blackmagic_hosted". 
PC-hosted semihosting does keyboard, file and screen i/o on the system where blackmagic_hosted runs, using linux system calls, while semihosting in the probe does keyboard, file and screen i/o on the system where gdb runs, using gdb file i/o calls.

The pc-hosted semihosting was tested with the sketches from the Arduino STM32duino-semihosting library. Output attached. 

[semihosting_testbench.txt](https://github.com/blacksphere/blackmagic/files/4642399/semihosting_testbench.txt)

### CODE REVIEW

**Comment by koendv** [comment]
"traceswo flush" patch implements the following feature:
When decoding traceswo, a write to stimulus channel 31 flushes the usb output buffer.

This allows implementing "flush()" by writing zeroes to channel 31.
The exact number of zeroes needed depends upon the SWO serial port input buffer. 64 is a good number.

File: src/target/cortexm.c [hunk 1]
```diff
@@ -1076,6 +1104,228 @@ static int cortexm_hostio_request(target *t)
...
+		if (fnam2_len == 0) break;
+		char *fnam2 = malloc(fnam2_len + 1);
+		if (fnam2 == NULL) break;
+		target_mem_read(t, fnam2, fnam2_taddr, fnam2_len + 1);
+		if (target_check_error(t)) break;
+		fnam2[fnam2_len]='\0';
+		ret = rename(fnam1, fnam2);
+		free(fnam1);
+		free(fnam2);
+		break;
```
Comment (TDHolmes) [inline src/target/cortexm.c:1254]: All of these error handling conditions after successful `malloc`'s fail to `free` the buffers before breaking. It looks like this is also an issue in some of the other commands

File: src/target/cortexm.c [hunk 2]
```diff
@@ -1076,6 +1104,228 @@ static int cortexm_hostio_request(target *t)
...
+		uint32_t buf_len = params[2];
+		if (buf_taddr == TARGET_NULL) break;
+		if (buf_len == 0) {ret = 0; break;}
+		uint8_t *buf = malloc(buf_len);
+		if (buf == NULL) break;
+		target_mem_read(t, buf, buf_taddr, buf_len);
+		if (target_check_error(t)) break;
+		ret = write(params[0] - 1, buf, buf_len);
+		free(buf);
+		//fflush(stdout);
```
Comment (TDHolmes) [inline src/target/cortexm.c:1189]: delete

File: src/target/cortexm.c (hunk 1 above)
Comment (koendv) [reply src/target/cortexm.c:1254]: You're right.

File: src/target/cortexm.c (hunk 2 above)
Comment (koendv) [reply src/target/cortexm.c:1189]: What do you wish to delete?

**Comment by TDHolmes** [comment]
Oh I just meant delete the commented out code

**Comment by koendv** [comment]
fixed, check again.

**Comment by TDHolmes** [comment]
`malloc`/`free` fix looks good! Can't really review the rest with much authority though, as I'm pretty new to all of this code.

**Comment by koendv** [comment]
On Mon, 25 May 2020 14:57:20 -0700
Tyler Holmes <notifications@github.com> wrote:

> `malloc`/`free` fix looks good! Can't really review the rest with
> much authority though, as I'm pretty new to all of this code.
> 

OK. It's the semihosting syscalls for the pc-based version of bmp.

**Comment by UweBonnes** [comment]
As stated on discord, please state explicit when you want me to merge.

**Comment by koendv** [comment]
On Tue, 26 May 2020 05:24:56 -0700
UweBonnes <notifications@github.com> wrote:

> As stated on discord, please state explicit when you want me to merge.
> 

OK. Please merge this PR. 

regards,

koen

**Comment by UweBonnes** [comment]
It is a good idea to rebase to master and run git log --check and git rebase --whitespace=fix master if problems show up.

**Comment by koendv** [comment]
ok. done.

**Comment by UweBonnes** [comment]
Can you please check that my multi branch has rebase your changes in the right way?

**Comment by koendv** [comment]
Source looks good. On Raspberry Raspbian "Buster": 
git clone --branch multi https://github.com/UweBonnes/blackmagic
make PROBE_HOST=hosted
LD_PRELOAD=/usr/lib/arm-linux-gnueabihf/libasan.so.5 ./src/blackmagic

Recognizes jlink, cmsis-dap, bmp.

**Pull request merged**